scripts/fast-reboot

```diff
@@ -70,44 +88,83 @@ docker exec -i bgp killall -9 bgpd
+    config warm_restart enable system
+fi
+# Kill swss dockers
+docker kill swss
+
 # syncd graceful stop is supported only for Broadcom platforms only for now
-if [[ "$sonic_asic_type" = 'broadcom' ]];
-then
+if [[ "$sonic_asic_type" = 'broadcom' ]]; then
     # Gracefully stop syncd
-    docker exec -i syncd /usr/bin/syncd_request_shutdown --cold > /dev/null
+    if [[ "$REBOOT_TYPE" = "fast-reboot" ]]; then
+        docker exec -i syncd /usr/bin/syncd_request_shutdown --cold > /dev/null
+    elif [[ "$REBOOT_TYPE" = "warm-reboot" ]]; then
+        docker exec -i syncd /usr/bin/syncd_request_shutdown --warm > /dev/null
+    fi
 
     # Check that syncd was stopped
     while docker top syncd | grep -q /usr/bin/syncd
     do
-          sleep 0.1
+        sleep 0.1
     done
 fi
 
+# Warm reboot: dump state to host disk
+# Note: even if syncd changed ASIC_DB before killed, we don't care
+if [[ "$REBOOT_TYPE" = "warm-reboot" ]]; then
+    # Dump redis content to directory
+    # Note: don't use pretty mode redis-dump (1.1) since there is a known bug with key pattern
+    DUMP_CMD="redis-dump -s /var/run/redis/redis.sock"
+    WARM_DIR=/host/warmboot
+    mkdir -p $WARM_DIR
+    # Note: requiring redis-dump-load
+    # Save applDB in /host/warm-reboot/appl_db.json
+    $DUMP_CMD -d 0 -o $WARM_DIR/appl_db.json
+    # Save configDB in /host/warm-reboot/config_db.json
+    $DUMP_CMD -d 4 -o $WARM_DIR/config_db.json
+    # Save stateDB (only FDB_TABLE and WARM_RESTART_TABLE) in /host/warm-reboot/state_db.json
+    # WARNING WARNING WARNING: a trick to dump both FDB_TABLE|* and WARM_RESTA*
+    # TODO: replace it with readable mechanism to dump multiple key patterns into one single json file
+    $DUMP_CMD -d 6 -k "[FW][DA][BR][_M][T_][AR][BE][LS][ET][|A]*" -o $WARM_DIR/state_db.json
+    # Save asicDB in /host/warm-reboot/asic_db.json
+    $DUMP_CMD -d 1 -o $WARM_DIR/asic_db.json
+fi
+
+# Kill syncd dockers
+docker kill syncd
+
 # Kill other containers to make the reboot faster
 docker ps -q | xargs docker kill > /dev/null
 
 # Stop the docker container engine. Otherwise we will have a broken docker storage
 systemctl stop docker.service
 
-# Stop opennsl modules for Broadcom platform
-if [[ "$sonic_asic_type" = 'broadcom' ]];
-then
-  service_name=$(systemctl list-units --plain --no-pager --no-legend --type=service | grep opennsl | cut -f 1 -d' ')
-  systemctl stop "$service_name"
-fi
-
-# Stop kernel modules for Nephos platform
-if [[ "$sonic_asic_type" = 'nephos' ]];
-then
-  systemctl stop nps-modules-`uname -r`.service
+if [[ "$sonic_asic_type" = 'broadcom' ]]; then
+    # Stop opennsl modules for Broadcom platform
+    service_name=$(systemctl list-units --plain --no-pager --no-legend --type=service | grep opennsl | cut -f 1 -d' ')
+    systemctl stop "$service_name"
+elif [[ "$sonic_asic_type" = 'nephos' ]]; then
+    # Stop kernel modules for Nephos platform
+    systemctl stop nps-modules-`uname -r`.service
```

Comment:
Do you think calculate service_name here and move 'systemctl stop "$service_name"' outside the if - elif -fi block would be better?

Reply:
Let's keep as is. Not in the scope of this PR.

---
In reply to: [225447840](https://github.com/Azure/sonic-utilities/pull/338#discussion_r225447840) [](ancestors = 225447840)